sshuttle/ssnet.py

```diff
@@ -703,14 +712,142 @@ def maybe_resume(self, wrote):
                 (_global_mux_wrapper_buffer_size, self.buf_total, self.channel))
 
 
+class Metrics:
+
+    def __init__(self):
+        self.max_handlers = 0  # Used to track the max number of handlers.
+        self.max_mux_wrapper_buffer_size = 0  # Used to track the max size of data to the edge.
+        self.paused_to_the_edge_count = 0  # Used to track the max size of data to the edge.
+        self.too_full_flush_count = 0  # Used to count the amount of times too_full was hit and mux flushed.
+        self.max_fullness = 0  # Used to track the max fullness of the mux in bytes.
+        self.max_outbufs = 0  # Used to track the max number of buffers seen for the mux.
+        self.dns_request_count = 0  # Used to count the number of dns request.
+        self.udp_request_count = 0  # Used to count the number of udp request.
+        self.max_dns_channels = 0  # Used to track the max number of channels being used by dns.
+        self.max_udp_channels = 0  # Used to track the max number of channels being used by udp.
+        self.executed_read_count = 0  # Used to track the number of sockets that were actually read.
+        self.executed_write_count = 0  # Used to track the number of sockets that were actually written too.
+        self.select_count = 0  # Used to track the number of time that select was called and processed the handlers.
+        self.time_to_send_metrics = time.time() + 60  # Send gauge metrics once a minute.
+
+    def log_metrics(self):
+        now = time.time()
+
+        if self.time_to_send_metrics < now:
+            self._log()
+            self.time_to_send_metrics = now + 60
+
+    def _log(self):
+        log('_log should be overloaded\n')
+
+    def save_max_handlers(self, handlers):
+        current_handlers = len(handlers)
```

Comment:
Is there any concern over the performance impact of calling the length function over these metrics? Assuming these save functions are called in the tight loop